### PR TITLE
Dynamic Dashboard: Update new cards notice and "Share Your Store" card behaviors

### PR DESCRIPTION
**Pull request by hafizrahman**
## Description

Based on discussion in Slack: p1717560988330199-slack-C03L1NF1EA3

**tl;dr:** We want the app to not show the new card notices ("Add More Sections" card and Edit button badge), even if there are new cards, if currently the app is showing the "Share Your Store" card. The rationale is that if that card is shown, then the store has no orders yet, and the new cards aren't too useful yet in this situation.

Additionally, I also updated the behavior for "Share Your Store" to not be shown by default, and also when loading. Instead, it will be shown once loading is done. It looks more neat that way, previously it could show up first and then disappeared after orders checking was done and there were orders found.

## Steps to reproduce

1. You will need two test sites: one with no order, and one with existing orders,
2. Start from logged out situation, then log in to the site with no order,
3. Once Dashboard is loaded, ensure that the "Share Your Store" card **is shown**, 
4. Ensure also that the Edit button **has no dot badge**, and the "Add More Sections" card **is not shown**,
5. Switch to the other test site with existing orders,
6. Once Dashboard is loaded, ensure that the "Share Your Store" card **is not shown**, 
7. Ensure also that the Edit button now **has a dot badge**, and the "Add More Sections" card **is shown**

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/208">19.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 1]
```diff
@@ -595,9 +595,10 @@ private extension DashboardViewModel {
     }
 
     /// Determines whether to show the notice that new cards now exist and can be found in Customize screen.
-    /// Can optionally pass local cards in case they are recently loaded before calling this function.
     func configureNewCardsNotice() {
-        guard featureFlagService.isFeatureFlagEnabled(.dynamicDashboardM2) else {
+        // If the site has no orders, the app will display the "Share Your Store" card.
+        // In this situation, it should keep the notice (both card and badge) hidden.
+        guard featureFlagService.isFeatureFlagEnabled(.dynamicDashboardM2) && hasOrders else {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:601]: I'm concerned about accessing `hasOrders` directly here. The reason is `configureNewCardsNotice` can be triggered by `updateDashboardCards`, which can be called from a Combine `sink` operator in `observeValuesForDashboardCards`.

The `@Published` properties that we observe in that operator emits new values just before the values are actually updated (`objectWillChange`). Accessing them directly while working with Combine can lead to undefined behaviors. The rule of thumb is when you work with Combine, ensure to inject parameters to helper methods instead of letting them access the properties directly. 

In this case, please consider injecting `hasOrders` to `configureNewCardsNotice` for safety.

**Review by itsmeichigo — APPROVED** [review]
This works as expected. However, I have a concern regarding accessing `hasOrders` from `configureNewCardsNotice` (more on the comment below).

I'm pre-approving this PR, please feel free to merge it after addressing the issue.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 1 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:601]: Good catch, thanks for the detailed information. Added with a bit of parameter refactoring in [becd61a](https://github.com/woocommerce/woocommerce-ios/pull/12985/commits/becd61adfa340511b2b3972d2621f1a58fefeb23)

**Comment by itsmeichigo** [comment]
Thanks, Hafiz. I'll merge this PR now.
